### DISCUSSION TITLE
Error with HeteroData nn captum explainer

### DISCUSSION THREAD

**Discussion by aihiangng** [pyg-team/pytorch_geometric#9758] [Q&A]
I followed the sample given here: https://github.com/pyg-team/pytorch_geometric/blob/master/torch_geometric/nn/models/captum.py#L326

```
from captum.attr import IntegratedGradients

from torch_geometric.data import HeteroData
from torch_geometric.nn import HeteroConv
from torch_geometric.nn import (captum_output_to_dicts,
                                to_captum_model, to_captum_input)

# data = HeteroData(...)
# model = HeteroConv(...)
...  # Train the model.

# Explain predictions for node `10`:
mask_type="edge"
metadata = data.metadata
output_idx = 10
captum_model = to_captum_model(model, mask_type, output_idx, metadata)
inputs, additional_forward_args = to_captum_input(data.x_dict,
                                    data.edge_index_dict, mask_type)

ig = IntegratedGradients(captum_model)
ig_attr = ig.attribute(inputs=inputs,
                       target=int(y[output_idx]),
                       additional_forward_args=additional_forward_args,
                       internal_batch_size=1)
edge_attr_dict = captum_output_to_dicts(ig_attr, mask_type, metadata)
```

And i get this error message. 

<img width="407" alt="image" src="https://github.com/user-attachments/assets/d5ceab7b-33d7-4927-87ed-b3a7f38f0b70">

<img width="381" alt="image" src="https://github.com/user-attachments/assets/15843262-85ed-4437-8add-85c973bad1bf">

**Comment by aihiangng** [comment]
ok solved it, should have been 
```
metadata = data.metadata() # add the brackets
```

**Comment by djdarcy** [comment]
## Related Projects & Prior Art

Some existing CLI-based AI assistant approaches worth studying:

### Microsoft AIShell
https://github.com/PowerShell/AIShell

Microsoft's AI-powered shell assistant. Similar concept to what we're proposing, though tightly integrated with PowerShell.

### OpenAI Codex CLI
https://github.com/openai/codex

OpenAI's terminal-based coding agent. More similar to Claude Code (AI-first with shell escape) than our proposed shell-first approach.

### Claude Code
https://claude.ai/code

Anthropic's CLI - the tool we're extending. AI-first paradigm with `!` to escape to shell.

---

## Vision Clarification: AISH as a Real Shell

The goal for AISH isn't just an AI wrapper - it should be a **real shell** that people could feasibly use as:

- **Linux login shell** - Replace bash/zsh as default
- **Windows CMD replacement** - Simpler than PowerShell, familiar syntax

### Why not PowerShell?

PowerShell is overkill for most tasks and has confusing syntax. Most of the time we just want to:
- Execute commands as we historically have
- Use familiar CMD/BASH patterns
- Get AI help *when needed*, not by default

### Design Principles

1. **CMD + BASH compatibility** - Familiar commands work as expected
2. **Minimal overhead** - Shell-first, AI on-demand (`~command #help`)
3. **Cross-platform** - Same experience on Windows/Linux/macOS
4. **Gradual AI integration** - Users opt-in to AI assistance per-command

This positions AISH as a practical daily-driver shell, not just a demo or wrapper.

---

See design doc: `2026-01-29__13-16-31__aish-dual-mode-shell-poc-design.md`
